### ISSUE TITLE
why would you caculate the upper triangle ?

### ISSUE THREAD

**Issue by ray-lee-94** [koshian2/affinity-loss#1]
I m interested in your implement of the affinity-loss. I try to covert it to the pytorch version. But i don't understand why would you caculate the upper triangle of the W. I need your help.

**Comment by ray-lee-94** [comment]
I rewrite the code as below:
```py

# affinity layer
class ClusteringAffinity(nn.Module):
    def __init__(self,n_classes,n_centers,sigma,feat_dim,init_weight=True,**kwargs):
        super(ClusteringAffinity,self).__init__()
        self.n_classes=n_classes
        self.n_centers=n_centers
        self.feat_dim=feat_dim
        self.sigma=sigma
        self.centers=nn.Parameter(torch.randn(self.n_classes,self.n_centers,self.feat_dim))
        if init_weight:
            self.__init_weight()

    def __init_weight(self):
        nn.init.kaiming_normal_(self.centers)
    def forward(self,f):
        f_expand=f.unsqueeze(1).unsqueeze(1)
        w_expand=self.centers.unsqueeze(0)
        fw_norm=torch.sum((f_expand-w_expand)**2,-1)
        distance=torch.exp(-fw_norm/self.sigma)
        distance=torch.max(distance,-1)[0]
#Regularization
        mc=self.n_centers*self.n_classes
        w_reshape=self.centers.view(mc,self.feat_dim)
        w_reshape_expand1=w_reshape.unsqueeze(0)
        w_reshape_expand2=w_reshape.unsqueeze(1)
        w_norm_mat=torch.sum((w_reshape_expand2-w_reshape_expand1)**2,-1)
        w_norm_upper=torch.triu(w_norm_mat)
        mu=2.0/(mc**2-mc)*w_norm_upper.sum()
        residuals=((w_norm_upper-mu)**2).triu()
        rw=2.0/(mc**2-mc)*residuals.sum()
        batch_size=f.size(0)
        rw_broadcast=torch.ones((batch_size,1)).to('cuda')*rw
        output=torch.cat((distance,rw_broadcast),dim=-1)
        return output
    def upper_triangle(self,metrix):
        pass
class Affinity_Loss(nn.Module):
    def __init__(self,lambd):
        super(Affinity_Loss,self).__init__()
        self.lamda=lambd
    def forward(self, y_pred_plusone,y_true_plusone):
        onehot=y_true_plusone[:,:-1]
        distance=y_pred_plusone[:,:-1]
        rw=torch.mean(y_pred_plusone[:,-1])
        d_fi_wyi=torch.sum(onehot*distance, -1).unsqueeze(1)
        losses=torch.clamp(self.lamda+distance-d_fi_wyi,min=0)
        L_mm=torch.sum(losses*(1-onehot))/y_true_plusone.size(0)
        return L_mm+rw
```
But the train loss  dose not drop a lot

**Comment by koshian2** [comment]
@VCBE123 
Let's look at the definition of diversity regularizer described [in the paper](https://arxiv.org/pdf/1901.07711.pdf).

<img src="https://latex.codecogs.com/gif.latex?\begin{align}&space;R(\boldsymbol{w})&=\mathbb{E}\bigl[(\|\boldsymbol{w}_j-\boldsymbol{w}_k\|^2-\mu)^2\bigr],&space;s.t.&space;j<k&space;\\&space;\mu&space;&=&space;\frac{2}{C^2-C}\sum_{j<k}\|\boldsymbol{w}_j-\boldsymbol{w}_k\|^2&space;\end{align}">

Where s.t. means a constraint condition that j <k.
Let's consider a 4 × 4 matrix written like this.

<img src="https://latex.codecogs.com/gif.latex?\begin{bmatrix}&space;a_{1,1}&space;&&space;a_{1,2}&space;&&space;a_{1,3}&space;&&space;a_{1,4}&space;\\&space;a_{2,1}&space;&&space;a_{2,2}&space;&&space;a_{2,3}&space;&&space;a_{2,4}&space;\\&space;a_{3,1}&space;&&space;a_{3,2}&space;&&space;a_{3,3}&space;&&space;a_{3,4}&space;\\&space;a_{4,1}&space;&&space;a_{4,2}&space;&&space;a_{4,3}&space;&&space;a_{4,4}&space;\end{bmatrix}">

Here, the upper triangular matrix without the diagonal component:

<img src="https://latex.codecogs.com/gif.latex?\begin{bmatrix}&space;0&space;&&space;a_{1,2}&space;&&space;a_{1,3}&space;&&space;a_{1,4}&space;\\&space;0&space;&&space;0&space;&&space;a_{2,3}&space;&&space;a_{2,4}&space;\\&space;0&space;&&space;0&space;&&space;0&space;&&space;a_{3,4}&space;\\&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;\end{bmatrix}">

The sum of this upper triangular matrix is ​​equal to

<img src="https://latex.codecogs.com/gif.latex?\sum_{j<k}a_{j,k}">

Thus, taking an upper triangular matrix is ​​necessary to satisfy the constraints in the definition of diversity regularizer. (Actual loss may not change much even if this condition is excluded.)
